src/lib.rs

```diff
@@ -0,0 +1,238 @@
+        })?;
+        rt.create(&st)?;
+        Ok(())
+    }
+
+    /// Records a pre-commitment from an actor to perform an atomic
+    /// execution. This method is to be invoked by a wrapped crossnet
+    /// message originating in one of the execution actors involved in
+    /// the atomic execution. Once the coordinator actor collects
+    /// pre-commitments from all the execution actors, it emits for
+    /// each of the execution actors a crossnet message triggering the
+    /// specified method to commit the atomic execution.
+    fn pre_commit<BS, RT>(rt: &mut RT, params: ApplyMsgParams) -> Result<bool, ActorError>
+    where
+        BS: Blockstore,
+        RT: Runtime<BS>,
+    {
+        let st: State = rt.state()?;
+        rt.validate_immediate_caller_is(std::iter::once(&st.ipc_gateway_address))?;
+
+        let ApplyMsgParams {
+            cross_msg:
+                CrossMsg {
+                    msg: StorableMsg { from, params, .. },
+                    ..
+                },
+        } = params;
+
+        let params: PreCommitParams = cbor::deserialize_params(&params)?;
+        let actors = &params.actors;
+        let exec_id = &params.exec_id;
+
+        if !actors.contains(&from) {
+            return Err(actor_error!(
+                illegal_argument,
+                "unexpected cross-message origin"
+            ));
+        }
+
+        let msgs = rt.transaction(|st: &mut State, rt| {
+            st.modify_atomic_exec(rt.store(), exec_id.into(), actors.clone(), |entry| {
+                // Record the pre-commitment
+                entry.set(BytesKey::from(from.to_bytes()?), params.commit)?;
+
+                // Check if any pre-commitment is missing
+                for actor in actors {
+                    let k = BytesKey::from(actor.to_bytes()?);
+                    if !entry.contains_key(&k)? {
+                        return Ok(None);
+                    }
+                }
+
+                // Prepare messages to commit the atomic execution
+                let mut msgs = Vec::new();
+                entry.for_each(|addr, &method| {
+                    msgs.push(CrossMsg {
+                        msg: StorableMsg {
+                            to: IPCAddress::from_bytes(addr)?,
+                            method,
+                            params: exec_id.to_bytes().into(),
+                            ..Default::default()
+                        },
+                        wrapped: true,
+                    });
+                    Ok(())
+                })?;
+                Ok(Some(msgs))
+            })
+            .map_err(|e| {
+                e.downcast_default(ExitCode::USR_ILLEGAL_STATE, "failed to update registry")
+            })
+        })?;
+
+        match msgs {
+            Some(msgs) => {
+                // Send the messages to commit the atomic execution
+                for msg in msgs {
+                    rt.send(
```

Comment:
Now that I see this call (and definitely future work), it may be a good idea to add to the gateway a `batchSend` method, WDYT? Let's add an issue if you feel this could be useful.

Reply:
We could make `SendCross` method always receive a set of messages, but simply supply a singleton set most of the time. WDYT?

Reply:
I like it. Do you mind tracking it in an issue in the IPC-gateway repo?